compiler/next/include/chpl/types/QualifiedType.h

```diff
@@ -86,14 +95,59 @@ class QualifiedType final {
     yet have a param value. */
   const Param* param() const { return param_; }
 
-  bool hasType() const {
+  bool hasTypePtr() const {
     return type_ != nullptr;
   }
-  bool hasParam() const {
+  bool hasParamPtr() const {
     return param_ != nullptr;
   }
 
-  bool isGenericOrUnknown() const;
+  // these are not defined here so we don't have to #include Type.h, Param.h
+  bool isGenericType() const;
+  bool isUnknownType() const;
+
+  bool isGeneric() const {
+    bool genericParam = kind_ == PARAM && !hasParamPtr();
+    return genericParam || isGenericType();
+  }
+
+  bool isUnknown() const {
+    return kind_ == UNKNOWN || !hasTypePtr() || isUnknownType();
+  }
+
+  bool isGenericOrUnknown() const { return isUnknown() || isGeneric(); }
+
+  /** Returns true if kind is TYPE */
+  bool isType() const { return kind_ == Kind::TYPE; }
+
+  /** Returns true if kind is PARAM */
+  bool isParam() const { return kind_ == Kind::PARAM; }
+
+  /**
+    Returns true if the value cannot be modified directly (but might
+    be modified by some other aliasing variable).
+    In particular, returns true for all kinds other than REF and VALUE.
+   */
+  bool isConst() const {
+    return kind_ == Kind::CONST_INTENT ||
+           kind_ == Kind::CONST_VAR ||
+           kind_ == Kind::CONST_REF ||
+           kind_ == Kind::CONST_IN ||
+           kind_ == Kind::PARAM ||
+           kind_ == Kind::TYPE ||
+           kind_ == Kind::FUNCTION ||
+           kind_ == Kind::MODULE;
+  }
```

Comment:
Are these QualifiedType classes also interned? And immutable? If so, could be have a bitset for each of these `isBlank` functions and compute them once on construction? (Relates to our discussion of comparing against interned strings)

Reply:
No, they're not interned. QualifiedType is mostly a pair of kind/intent and a type pointer. The type pointer is interned/managed by the context.

Anyway I'm not sure why having a bitset would help (these are probably fast enough to compute, memory access probably a bigger deal than comparing against a bunch of constants).